fendermint/vm/interpreter/src/fvm/state/ipc.rs

```diff
@@ -263,16 +275,50 @@ impl<DB: Blockstore> GatewayCaller<DB> {
         &self,
         state: &mut FvmExecState<DB>,
         cross_messages: Vec<IpcEnvelope>,
-    ) -> anyhow::Result<FvmApplyRet> {
-        let messages = cross_messages
-            .into_iter()
-            .map(xnet_messaging_facet::IpcEnvelope::try_from)
-            .collect::<Result<Vec<_>, _>>()
-            .context("failed to convert cross messages")?;
-        let r = self
+        final_ret: &mut TopDownApplyRetAggregator,
+    ) -> anyhow::Result<()> {
+        for msg in cross_messages {
+            let msg = xnet_messaging_facet::IpcEnvelope::try_from(msg)
+                .context("failed to convert cross messages")?;
+            self.apply_cross_msg(state, msg, final_ret)?;
+        }
+
+        Ok(())
+    }
+
+    fn apply_cross_msg(
+        &self,
+        state: &mut FvmExecState<DB>,
+        msg: xnet_messaging_facet::IpcEnvelope,
+        final_ret: &mut TopDownApplyRetAggregator,
+    ) -> anyhow::Result<()> {
+        let (msg, success, ret) = match self
+            .xnet
+            .call_with_return(state, |c| c.apply_msg_with_ret(msg.clone()))
```

Comment:
If you have to clone here anyway so that you can still access `msg` in the error case, then I really don't see why `cross_msg` is returned with the result. Can it be different from the input?